libs/core/langchain_core/memory.py

```diff
@@ -81,3 +89,158 @@ def clear(self) -> None:
     async def aclear(self) -> None:
         """Clear memory contents."""
         await run_in_executor(None, self.clear)
+
+
+class ChatMessageHistory(BaseChatMessageHistory, BaseModel):
+    """In memory implementation of chat message history.
+
+    Stores messages in an in memory list.
+    """
+
+    messages: List[BaseMessage] = Field(default_factory=list)
+
+    async def aget_messages(self) -> List[BaseMessage]:
+        return self.messages
+
+    def add_message(self, message: BaseMessage) -> None:
+        """Add a self-created message to the store"""
+        self.messages.append(message)
+
+    async def aadd_messages(self, messages: Sequence[BaseMessage]) -> None:
+        """Add messages to the store"""
+        self.add_messages(messages)
+
+    def clear(self) -> None:
+        self.messages = []
+
+    async def aclear(self) -> None:
+        self.clear()
+
+
+class BaseChatMemory(BaseMemory, ABC):
```

Comment:
@leo-gan we don't have a clear road map for the old memory abstractions yet, so I'm not sure what packages make sense for the abstractions

The `BaseMemory`  abstraction is problematic -- overly complex and incorrect interface. We'll likely end up deprecating it longer-term.

It could potentially go in langchain_community, but not sure what we'd want to do with `ConversationBufferMemory`.

Overall dependency goals are:

langchain-core: base abstractions and core implementations
langchain: implementations of common utilities (e.g., token counting, generic agent implements etc)
community / partner packages: specific implementations of the abstractions

So:
* langchain-core without dependencies
* langchain depends on langchain-core, it does not depend on langchain-community
* langchain-community depends on langchain-core and langchain
* partner packaegs depending on langchain-core (and maybe langchain)

---

cc @efriis  / @baskaryan